### PR TITLE
multi bls key support

### PR DESCRIPTION
**Pull request by gupadhyaya**
- [testnet] use a new testnet for multi-key test

Signed-off-by: Leo Chen <leo@harmony.one>

- adding support for multiple initial accounts for multi-key scenarios

- minor fixes

- changing viewchange duration to exponential

[multi-key] use 2 shards for testnet of multi-key (#2)

Signed-off-by: Leo Chen <leo@harmony.one>

### Unit Test Coverage

Before:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

After:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **YES|NO**

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

3. **Describe how the plan was tested.**

4. **How much minimum baking period after the last flag epoch should we allow on Pangaea before promotion onto mainnet?**

5. **What are the planned flag epoch numbers and their ETAs on Pangaea?**

6. **What are the planned flag epoch numbers and their ETAs on mainnet?**

    Note that this must be enough to cover baking period on Pangaea.

7. **What should node operators know about this planned change?**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **YES|NO**

9. **Does the existing `node.sh` continue to work with this change?**

10. **What should node operators know about this change?**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

### CODE REVIEW

**Comment by gupadhyaya** [comment]
@rlan35 @fxfactorial @LeoHChen when you get chance, please review this PR, so that we can make progress. Fixed the bug that was spoiling shard chain explorer. The multi-bls network is running fine. I will continue to test the cross-shard tx.

File: cmd/harmony/main.go [hunk 1]
```diff
@@ -116,8 +115,10 @@ var (
 	enableMemProfiling = flag.Bool("enableMemProfiling", false, "Enable memsize logging.")
 	enableGC           = flag.Bool("enableGC", true, "Enable calling garbage collector manually .")
 	blsKeyFile         = flag.String("blskey_file", "", "The encrypted file of bls serialized private key by passphrase.")
-	blsPass            = flag.String("blspass", "", "The file containing passphrase to decrypt the encrypted bls file.")
-	blsPassphrase      string
+	blsFolder          = flag.String("blsfolder", "", "The encrypted file of bls serialized private key by passphrase.")
```
Comment (rlan35) [inline cmd/harmony/main.go]: description not updated.

File: cmd/harmony/main.go [hunk 2]
```diff
@@ -214,59 +219,103 @@ func passphraseForBls() {
 	blsPassphrase = passphrase
 }
 
-func setupInitialAccount() (isLeader bool) {
+func findAccountsByPubKeys(config shardingconfig.Instance, pubKeys []*bls.PublicKey) {
+	for _, key := range pubKeys {
+		_, account := config.FindAccount(key.SerializeToHexStr())
+		if account != nil {
+			initialAccounts = append(initialAccounts, account)
+		}
```
Comment (rlan35) [inline cmd/harmony/main.go]: let's printout if a key is not found for informational purpose.

File: consensus/consensus_leader_msg.go [hunk 3]
```diff
@@ -42,7 +42,7 @@ func (consensus *Consensus) constructPreparedMessage() ([]byte, *bls.Sign) {
 	}
 
 	consensusMsg := message.GetConsensus()
-	consensus.populateMessageFields(consensusMsg)
+	consensus.populateMessageFields(consensusMsg, consensus.LeaderPubKey)
```
Comment (rlan35) [inline consensus/consensus_leader_msg.go]: Why put as a separate field. Isn't populateMessageFields already have access to consensus.LeaderPubKey?

File: consensus/consensus_v2.go [hunk 4]
```diff
@@ -290,16 +297,19 @@ func (consensus *Consensus) onAnnounce(msg *msg_pb.Message) {
...
-		utils.Logger().Info().
-			Str("blockHash", hex.EncodeToString(consensus.blockHash[:])).
-			Msg("[OnAnnounce] Sent Prepare Message!!")
+	groupID := []nodeconfig.GroupID{nodeconfig.NewGroupIDByShardID(nodeconfig.ShardID(consensus.ShardID))}
+	for i, key := range consensus.PubKey.PublicKey {
+		// Construct and send prepare message
+		msgToSend := consensus.constructPrepareMessage(key, consensus.priKey.PrivateKey[i])
+		// TODO: this will not return immediatey, may block
+
+		if err := consensus.msgSender.SendWithoutRetry(groupID, host.ConstructP2pMessage(byte(17), msgToSend)); err != nil {
```
Comment (rlan35) [inline consensus/consensus_v2.go]: So we are not grouping the messages together in a single message?

File: consensus/consensus_v2.go [hunk 5]
```diff
@@ -103,7 +103,12 @@ func (consensus *Consensus) announce(block *types.Block) {
 
 	consensus.block = encodedBlock
 	consensus.blockHeader = encodedBlockHeader
-	msgToSend := consensus.constructAnnounceMessage()
+	key, err := consensus.GetConsensusLeaderPrivateKey()
```
Comment (rlan35) [inline consensus/consensus_v2.go]: This call is a bit expensive. Why can't we store the key in a field and directly read it? It's doing the looping everytime.

File: cmd/harmony/main.go [hunk 6]
```diff
@@ -311,7 +360,11 @@ func setupConsensusAndNode(nodeConfig *nodeconfig.ConfigType) *node.Node {
 	currentConsensus, err := consensus.New(
 		myHost, nodeConfig.ShardID, p2p.Peer{}, nodeConfig.ConsensusPriKey, decider,
 	)
-	currentConsensus.SelfAddress = common.ParseAddr(initialAccount.Address)
+
+	currentConsensus.SelfAddress = make(map[string]ethCommon.Address)
+	for _, initialAccount := range initialAccounts {
```
Comment (fxfactorial) [inline cmd/harmony/main.go]: leave def of selfaddress alone, instead add new field of  a map[shard.BlsPublicKey]struct{}

**Comment by fxfactorial** [comment]
So I have a suggestion for a different approach because the current approach will mean reimplenting again on master because otherwise the merge conflicts will be insane. 

Instead consider this. 

You make some kind of interface, like so: 

```go
type BLSSigner interface {
    SignForKeyIndex(index int, payload []byte) *bls.Sign
}
```

or something like that, and you attach a concrete type implementation of this interface to the consensus record itself, like how Decider is done. Now there are no function signatures to change, just use this field when need to sign. Adding this kind of thing will be much easier to master and will make it easy later for if we want to change the bls key management impl.

File: cmd/harmony/main.go (hunk 1 above)
Comment (LeoHChen) [reply cmd/harmony/main.go]: also, please add a message saying that all blskey has to be in the same shard, also share the same pass.

File: cmd/harmony/main.go (hunk 2 above)
Comment (LeoHChen) [reply cmd/harmony/main.go]: ditto

File: cmd/harmony/main.go [hunk 7]
```diff
@@ -214,59 +219,103 @@ func passphraseForBls() {
...
+	for _, account := range initialAccounts {
+		fmt.Printf("My Genesis Account: %v\n", *account)
+	}
 }
 
-func setupConsensusKey(nodeConfig *nodeconfig.ConfigType) *bls.PublicKey {
-	consensusPriKey, err := blsgen.LoadBlsKeyWithPassPhrase(*blsKeyFile, blsPassphrase)
+func readMultiBlsKeys(consensusMultiBlsPriKey *nodeconfig.MultiBlsPrivateKey, consensusMultiBlsPubKey *nodeconfig.MultiBlsPublicKey) error {
+	multiBlsKeyDir := blsFolder
+	blsKeyFiles, err := ioutil.ReadDir(*multiBlsKeyDir)
```
Comment (LeoHChen) [inline cmd/harmony/main.go]: please do a better job to read only .key files.

File: consensus/view_change.go [hunk 8]
```diff
@@ -121,20 +121,22 @@ func (consensus *Consensus) startViewChange(viewID uint64) {
 	consensus.current.SetViewID(viewID)
 	consensus.LeaderPubKey = consensus.GetNextLeaderKey()
 
-	diff := viewID - consensus.viewID
-	duration := time.Duration(int64(diff) * int64(viewChangeDuration))
+	diff := int64(viewID - consensus.viewID)
+	duration := time.Duration(diff * diff * int64(viewChangeDuration))
```
Comment (LeoHChen) [inline consensus/view_change.go]: this change has to create a separate PR.

File: internal/configs/sharding/testnet.go [hunk 9]
```diff
@@ -16,9 +16,9 @@ const (
 	testnetVdfDifficulty = 10000 // This takes about 20s to finish the vdf
 
 	// TestNetHTTPPattern is the http pattern for testnet.
-	TestNetHTTPPattern = "https://api.s%d.b.hmny.io"
+	TestNetHTTPPattern = "https://api.s%d.b2.hmny.io"
```
Comment (LeoHChen) [inline internal/configs/sharding/testnet.go]: this change b2 shouldn't be here.

File: cmd/harmony/main.go (hunk 1 above)
Comment (gupadhyaya) [reply cmd/harmony/main.go]: `blsFolder          = flag.String("blsfolder", ".hmy/blskeys", "The folder that stores the bls keys; same blspass is used to decrypt all bls keys; all bls keys mapped to same shard")`

File: cmd/harmony/main.go (hunk 2 above)
Comment (gupadhyaya) [reply cmd/harmony/main.go]: `fmt.Printf("Bls key not found: %s\n", keyStr)`

File: cmd/harmony/main.go (hunk 7 above)
Comment (gupadhyaya) [reply cmd/harmony/main.go]: `
if filepath.Ext(blsKeyFile.Name()) != ".key" {
			continue
		}
`

File: consensus/consensus_leader_msg.go (hunk 3 above)
Comment (gupadhyaya) [reply consensus/consensus_leader_msg.go]: Also used by non-leader during `constructPrepareMessage` which needs non-leader pub key to be populated.

File: consensus/consensus_v2.go (hunk 4 above)
Comment (gupadhyaya) [reply consensus/consensus_v2.go]: Should be possible, but requires more changes to signing and verifying. I was planning to consider this for the open staking multi key support.

File: consensus/consensus_v2.go (hunk 5 above)
Comment (gupadhyaya) [reply consensus/consensus_v2.go]: Agreed. I will add this optimization to the open staking multi-key support.

File: cmd/harmony/main.go (hunk 6 above)
Comment (gupadhyaya) [reply cmd/harmony/main.go]: why do we need two separate fields? The only use of the SelfAddress is to correctly assign the address of the block proposer while proposing the block. 
`coinbase := node.Consensus.SelfAddress[node.Consensus.LeaderPubKey.SerializeToHexStr()]` 
SelfAddress is not used anywhere else.

**Comment by gupadhyaya** [comment]
> So I have a suggestion for a different approach because the current approach will mean reimplenting again on master because otherwise the merge conflicts will be insane.
> 
> Instead consider this.
> 
> You make some kind of interface, like so:
> 
> ```go
> type BLSSigner interface {
>     SignForKeyIndex(index int, payload []byte) *bls.Sign
> }
> ```
> 
> or something like that, and you attach a concrete type implementation of this interface to the consensus record itself, like how Decider is done, the dependency injection way. Now there are no function signatures to change, just use this field when need to sign. Adding this kind of thing will be much easier to master and will make it easy later for if we want to change the bls key management impl.

Great suggestion. But, I want to consider this for the master only. Anyways, we will have to discard these changes when we move to master (as master has deviated quite a bit from current s3). Do you agree?

**Comment by fxfactorial** [comment]
No, that's the point. By writing behind an interface and your concrete type, will be less code to throw away since you won't touch function signatures by by way suggested and the work will have less chance of merge conflict with master if you try a cherry pick with work already done.

File: cmd/harmony/main.go (hunk 6 above)
Comment (fxfactorial) [reply cmd/harmony/main.go]: To lessen the pain of merge conflicts to master

**Pull request merged**